packages/flutter_tools/templates/ffi_plugin/lib/projectName.dart.tmpl

```diff
@@ -0,0 +1,70 @@
+{{#no_platforms}}
+// You have generated a new plugin project without
+// specifying the `--platforms` flag. A plugin project supports no platforms is generated.
+// To add platforms, run `flutter create -t plugin_ffi --platforms <platforms> .` under the same
+// directory. You can also find a detailed instruction on how to add platforms in the `pubspec.yaml` at https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin-platforms.
+{{/no_platforms}}
+
+import 'dart:async';
+import 'dart:ffi';
+import 'dart:io';
+import 'dart:isolate';
+
+const libName = '{{projectName}}';
+
+final DynamicLibrary _dylib = () {
+  if (Platform.isMacOS || Platform.isIOS) {
+    return DynamicLibrary.open('$libName.framework/$libName');
+  }
+  if (Platform.isAndroid || Platform.isLinux) {
+    return DynamicLibrary.open('lib$libName.so');
+  }
+  if (Platform.isWindows) {
+    return DynamicLibrary.open('$libName.dll');
+  }
+  throw UnsupportedError('Unimplemented!');
+}();
+
+final sum = _dylib.lookupFunction<IntPtr Function(IntPtr, IntPtr),
+    int Function(int, int)>('sum');
+
+bool _initializedDartApiDL = false;
+void _ensureInitializedDartApiDL() {
+  if (_initializedDartApiDL) {
+    return;
+  }
+  final initializeApi = _dylib.lookupFunction<IntPtr Function(Pointer<Void>),
+      int Function(Pointer<Void>)>("Dart_InitializeApiDL");
+  final initializeResult = initializeApi(NativeApi.initializeApiDLData);
+  if (initializeResult != 0) {
+    throw "Failed to initialize API.";
+  }
+  if (_dylib.lookup<IntPtr>("Dart_PostCObject_DL").value == 0) {
+    throw "Failed to initialize API.";
+  }
+}
+
+final _sumAsync = () {
+  _ensureInitializedDartApiDL();
+  return _dylib.lookupFunction<IntPtr Function(Int64 port, IntPtr, IntPtr),
+      int Function(int port, int, int)>('sum_async');
+}();
+
+Future<int> sumAsync(int a, int b) async {
+  final completer = Completer<int>();
+  // Note: if this an often called function, it would be better to reuse the
+  // native port rather than creating a new one on every call.
```

Comment:
Is it ever better to _not_ reuse the native port? I guess if it's infrequently called you save memory, but do we know how much? 

Also leaving ports open can cause your application to hang/not exit gracefully.

Reply:
We have multiple possible scenarios to put in the template:

1. An infrequent Dart FFI call that leads to a single asynchronous callback
    * Dart API is `Future`.
    * Invoking the Dart closure opens the port.
    * The async callback closes the port.
    * The port is the identity that connects the call with the callback.
2. An infrequent Dart FFI call that leads to multiple asynchronous callbacks
    * Dart API is `Stream`.
    * Invoking the Dart closure opens the port.
    * The last callback should call the `close` method on the stream.
    * The port is the identity that connects the call with the callback.
3. A frequent call, single callback, and the native code is guaranteed FIFO
    * Dart API is `Future`.
    * Invoking the Dart closure opens a port if the previous port has been closed.
    * We need a close method somewhere that closes the port.
    * The async callbacks in Dart complete `Future`s in order.
4. A frequent call, single callback, and the native function passes through a nonce.
    * Dart API is `Future`.
    * Invoking the Dart closure opens a port if the previous port has been closed.
    * We need a close method somewhere that closes the port.
    * The Dart FFI call passes a nonce to C, which passes it back in the async callback to know which Future to complete.
5. ..., etc.

It really depends on the use case what the (1) Dart API is, (2) if a port is kept open and/or reused for multiple calls and callbacks, and (3) if the native code guarantees FIFO and/or allows passing nonces.

So the question is what example we should put in the sample, or should we put in multiple?

Reply:
It might be good to just have more documentation in the sample about what to do depending on what kind of operation you are writing.